File: bothan-api/client/go-client/rest.go
package client

import (
	"encoding/json"
	"net/url"
	"path"
	"strings"
	"time"

	proto "github.com/bandprotocol/bothan/bothan-api/client/go-client/proto/bothan/v1"
	"github.com/levigross/grequests"
)

var _ Client = &RestClient{}

type RestClient struct {
	url     string
	timeout time.Duration
}

func NewRestClient(url string, timeout time.Duration) *RestClient {
	return &RestClient{url, timeout}
}

func (c *RestClient) GetInfo() (*proto.GetInfoResponse, error) {
	parsedUrl, err := url.Parse(c.url + "/info")
	if err != nil {
		return nil, err
	}

	resp, err := grequests.Get(
		parsedUrl.String(),
		&grequests.RequestOptions{
			RequestTimeout: c.timeout,
		},
	)
	if err != nil {
		return nil, err
	}

	if !resp.Ok {
		return nil, resp.Error
	}

	var infoResp proto.GetInfoResponse
	err = json.Unmarshal(resp.Bytes(), &infoResp)
	if err != nil {
		return nil, err
	}

	return &infoResp, nil
}

func (c *RestClient) UpdateRegistry(ipfsHash string, version string) error {
	parsedUrl, err := url.Parse(c.url + "/registry")
	if err != nil {
		return err
	}

	resp, err := grequests.Post(
		parsedUrl.String(), &grequests.RequestOptions{
			RequestTimeout: c.timeout,
			JSON: map[string]string{
				"ipfsHash": ipfsHash,
				"version":  version,
			},
		},
	)
	if err != nil {
		return err
	}

	if !resp.Ok {
		return resp.Error
	}

	return nil
}

func (c *RestClient) PushMonitoringRecords(uuid, txHash string, signalIDs []string) error {
	parsedUrl, err := url.Parse(c.url + "/monitoring_records")
	if err != nil {
		return err
	}

	resp, err := grequests.Post(
		parsedUrl.String(), &grequests.RequestOptions{
			RequestTimeout: c.timeout,
			JSON: map[string]any{
				"uuid":       uuid,
				"tx_hash":    txHash,
				"signal_ids": signalIDs,
			},
		},
	)

	if err != nil {
		return err
	}

	if !resp.Ok {
		return resp.Error
	}

	return nil
}

func (c *RestClient) GetPrices(signalIDs []string) (*proto.GetPricesResponse, error) {
	parsedUrl, err := url.Parse(c.url + "/prices")
	if err != nil {
		return nil, err
	}
	parsedUrl.Path = path.Join(parsedUrl.Path, strings.Join(signalIDs, ","))

	resp, err := grequests.Get(
		parsedUrl.String(),
		&grequests.RequestOptions{
			RequestTimeout: c.timeout,
		},
	)
	if err != nil {
		return nil, err
	}

	if !resp.Ok {
		return nil, resp.Error
	}

	var priceResp proto.GetPricesResponse
	err = json.Unmarshal(resp.Bytes(), &priceResp)
	if err != nil {
		return nil, err
	}

	return &priceResp, nil
}

File: bothan-api/client/rust-client/src/client/rest.rs
use reqwest::{Client, Error};
use url::{ParseError, Url};

use crate::proto::bothan::v1::{GetInfoResponse, GetPricesResponse};

pub struct RestClient {
    url: Url,
    client: Client,
}

impl RestClient {
    pub fn new(url: String) -> Result<Self, ParseError> {
        Ok(RestClient {
            url: Url::parse(&url)?,
            client: Client::new(),
        })
    }

    pub async fn get_info(&self) -> Result<GetInfoResponse, Error> {
        let mut url = self.url.clone();
        url.set_path("/info");
        let response = self.client.get(url).send().await?.error_for_status()?;
        let get_info_response = response.json().await?;
        Ok(get_info_response)
    }

    pub async fn update_registry(&self, ipfs_hash: &str, version: &str) -> Result<(), Error> {
        let mut url = self.url.clone();
        url.set_path("/registry");
        let payload = serde_json::json!({ "ipfs_hash": ipfs_hash, "version": version });
        let _ = self
            .client
            .post(url)
            .json(&payload)
            .send()
            .await?
            .error_for_status()?;
        Ok(())
    }

    pub async fn push_monitoring_records(
        &self,
        uuid: &str,
        tx_hash: &str,
        signal_ids: &[&str],
    ) -> Result<(), Error> {
        let mut url = self.url.clone();
        url.set_path("/monitoring_records");
        let payload =
            serde_json::json!({ "uuid": uuid, "tx_hash": tx_hash, "signal_ids": signal_ids });
        let _ = self
            .client
            .post(url)
            .json(&payload)
            .send()
            .await?
            .error_for_status()?;
        Ok(())
    }

    pub async fn get_prices(&self, signal_ids: &[&str]) -> Result<GetPricesResponse, Error> {
        let mut url = self.url.clone();
        url.set_path("/prices");
        let payload = serde_json::json!({ "signal_ids": signal_ids });
        let response = self
            .client
            .post(url)
            .json(&payload)
            .send()
            .await?
            .error_for_status()?;
        let get_prices_response = response.json().await?;
        Ok(get_prices_response)
    }
}
